### PR TITLE
Add container_tests for puller and Update load and pull rules to comply with updates in #946

### PR DESCRIPTION
**Pull request by xwinxu**
Change the BUILD file generated to make a container_import call.

Resolve conflict with container_import already creating a .tar output.

In our testing:
- We check content of symlinks
- We check 3 cases in container_tests:
          1) Presence of OCI image format only
          2) Presence of image.tar via container_import target
          3) Presence of image.tar via "docker" format option

### CODE REVIEW

File: tests/docker/BUILD [hunk 1]
```diff
@@ -390,35 +390,35 @@ file_test(
       }
    ]
 }""",
-    file = "@new_distroless_base//image:index.json",
+    file = "@new_distroless_base//image-oci:index.json",
```
Comment (xingao267) [inline tests/docker/BUILD]: Should also test the symlinks in the image directory.

File: container/new_pull.bzl [hunk 2]
```diff
@@ -91,21 +91,28 @@ def _impl(repository_ctx):
...
     name = "image",
-    srcs = glob(["image/**"]),
+    config = "config.json",
+    layers = glob(["*.tar.gz"]),
 )
 
+exports_files(glob(["**"]))""")
+
+    # Currently exports all files pulled by the binary and will not be depended on by other rules_docker rules.
+    repository_ctx.file("image-oci/BUILD", """package(default_visibility = ["//visibility:public"])
```
Comment (xingao267) [inline container/new_pull.bzl]: r/

File: container/new_pull.bzl [hunk 3]
```diff
@@ -91,21 +91,28 @@ def _impl(repository_ctx):
...
 
-filegroup(
+container_import(
     name = "image",
-    srcs = glob(["image/**"]),
+    config = "config.json",
+    layers = glob(["*.tar.gz"]),
 )
 
+exports_files(glob(["**"]))""")
```
Comment (xingao267) [inline container/new_pull.bzl]: We probably just need to export the image.tar file now.

File: container/new_pull.bzl [hunk 4]
```diff
@@ -91,21 +91,28 @@ def _impl(repository_ctx):
     # Add an empty top-level BUILD file.
     repository_ctx.file("BUILD", "")
 
-    # Currently exports all files pulled by the binary and cannot be depended on by other rules_docker rules.
-    # We will implement a new_container_import rule to comprehend this oci layout.
+    # Create symlinks to the appropriate config.json and layers in the pulled OCI image in the image-oci directory.
+    # Generates container_import rule which is able to comprehend OCI layout via the symlinks.
```
Comment (xingao267) [inline container/new_pull.bzl]: Generates container_import target ...

File: tests/docker/BUILD (hunk 1 above)
Comment (xwinxu) [reply tests/docker/BUILD]: added. I also added some for the loader too

File: container/new_load.bzl [hunk 5]
```diff
@@ -25,21 +25,24 @@ def _impl(repository_ctx):
...
+container_import(
     name = "image",
-    srcs = glob(["image/**"]),
+    config = "config.json",
+    layers = glob(["*.tar.gz"]),
 )
-exports_files(glob(["**"]))
-""", executable = False)
+
+exports_files(glob(["**"]))""")
```
Comment (xingao267) [inline container/new_load.bzl]: You might not need to export all files in this case.

File: container/new_pull.bzl [hunk 6]
```diff
@@ -91,21 +90,30 @@ def _impl(repository_ctx):
     # Add an empty top-level BUILD file.
     repository_ctx.file("BUILD", "")
 
-    # Currently exports all files pulled by the binary and cannot be depended on by other rules_docker rules.
-    # We will implement a new_container_import rule to comprehend this oci layout.
-    repository_ctx.file("image/BUILD", """package(default_visibility = ["//visibility:public"])
+    # Create symlinks to the appropriate config.json and layers in the pulled OCI image in the image-oci directory.
+    # Generates container_import rule which is able to comprehend OCI layout via the symlinks.
+    if repository_ctx.attr.format != "docker":
```
Comment (xingao267) [inline container/new_pull.bzl]: I would probably do

if repository_ctx.attr.format == "docker":
   ...
else
  ...

File: testdata/BUILD [hunk 7]
```diff
@@ -212,6 +214,25 @@ container_image(
     mode = "0o644",
 )
 
+# Test new_container_load
+container_image(
+    name = "new_docker_tarball_base",
```
Comment (xingao267) [inline testdata/BUILD]: new_docker_tarball_base suggests it's using a tarball as base. But new_pause_tar seems pull the image as oci.

File: testing/examples/WORKSPACE [hunk 8]
```diff
@@ -61,6 +62,49 @@ pip_import(
     requirements = "@base_images_docker//package_managers:requirements-pip.txt",
 )
 
+new_container_pull(
```
Comment (xingao267) [inline testing/examples/WORKSPACE]: Add these in the top level WORKSPACE. Targets here are not regularly tested.

File: testing/examples/basic/BUILD [hunk 9]
```diff
@@ -36,3 +36,43 @@ container_test(
     driver = "tar",
     image = "@basic_alpine_dockerfile//image:dockerfile_image",
 )
+
+# Tests for new_container_pull
```
Comment (xingao267) [inline testing/examples/basic/BUILD]: Add to tests/container/BUILD

File: testing/examples/extended/BUILD [hunk 10]
```diff
@@ -64,3 +64,87 @@ container_test(
     driver = "tar",
     image = "@extended_alpine_dockerfile//image:dockerfile_image",
 )
+
+# Test the OCI image option of new_container_pull
+container_image(
```
Comment (xingao267) [inline testing/examples/extended/BUILD]: Not need to add examples in testing/examples. Just the basic cases in tests/container/BUILD.

File: tests/container/BUILD [hunk 11]
```diff
@@ -416,18 +416,34 @@ file_test(
...
 }""",
-    file = "@new_distroless_base_both//image:oci-layout",
+    file = "@new_alpine_linux_armv6//image-oci:oci-layout",
+)
+
+# Testing for presence of symlinks
+file_test(
+    name = "new_distroless_base_oci_test_config_symlink",
+    file = "@new_alpine_linux_armv6//image:config.json",
+    regexp = ".*",
```
Comment (xingao267) [inline tests/container/BUILD]: Is it possible to test the content of the symlink?

File: tests/container/BUILD [hunk 12]
```diff
@@ -416,18 +416,34 @@ file_test(
...
+)
+
+file_test(
+    name = "new_distroless_base_oci_test_1targz_symlink",
```
Comment (xingao267) [inline tests/container/BUILD]: new_distroless_base_oci_test_targz_symlink

File: testdata/BUILD [hunk 13]
```diff
@@ -54,6 +54,8 @@ package(default_visibility = ["//visibility:public"])
 exports_files([
     "pause.tar",
     "go_pusher_test.tar",
+    "bar",
```
Comment (xingao267) [inline testdata/BUILD]: Why do you need to export these files?

File: WORKSPACE [hunk 14]
```diff
@@ -109,11 +162,28 @@ load(
...
+# Images for testing new_container_load
+new_container_pull(
+    name = "new_ubuntu1604",
+    registry = "l.gcr.io",
+    repository = "google/ubuntu1604",
+    tag = "latest",
+)
+
+new_container_pull(
+    name = "new_ubuntu1604_tar",
```
Comment (xingao267) [inline WORKSPACE]: Is it used?

File: WORKSPACE [hunk 15]
```diff
@@ -109,11 +162,28 @@ load(
...
 )
 
+# To test the new_container_load rule.
 new_container_load(
     name = "new_pause_tar",
     file = "//testdata:pause.tar",
 )
 
+# Images for testing new_container_load
+new_container_pull(
```
Comment (xingao267) [inline WORKSPACE]: Is it used?

Comment (xingao267) [reply WORKSPACE]: Saw that in the testdata/BUILD. But this is image is big, we probably don't want to have it here.

File: testing/examples/WORKSPACE [hunk 16]
```diff
@@ -61,6 +62,13 @@ pip_import(
...
+    name = "new_alpine_linux_amd64",
```
Comment (xingao267) [inline testing/examples/WORKSPACE]: You can just name it as alpine_linux_amd64.

File: testing/examples/WORKSPACE [hunk 17]
```diff
@@ -61,6 +62,13 @@ pip_import(
...
+    registry = "index.docker.io",
+    repository = "library/alpine",
+    tag = "3.8",
+)
+
 container_pull(
```
Comment (xingao267) [inline testing/examples/WORKSPACE]: Change all the other container_pull here to new_container_pull.

File: tests/container/BUILD [hunk 18]
```diff
@@ -443,15 +593,33 @@ file_test(
...
     "imageLayoutVersion": "1.0.0"
 }""",
-    file = "@new_pause_tar//image:oci-layout",
+    file = "@new_pause_tar//image-oci:oci-layout",
+)
+
+file_test(
+    name = "new_pause_tar_test_config_symlink",
+    file = "@new_pause_tar//image:config.json",
+    regexp = ".*",
```
Comment (xingao267) [inline tests/container/BUILD]: Add content of the config.json?

File: WORKSPACE [hunk 19]
```diff
@@ -56,26 +56,79 @@ new_container_pull(
...
+)
+
+new_container_pull(
+    name = "new_distroless_fixed_id",
+    digest = "sha256:a26dde6863dd8b0417d7060c990abe85c1d2481541568445e82b46de9452cf0c",
+    registry = "gcr.io",
+    repository = "distroless/base",
+)
+
+new_container_pull(
```
Comment (xingao267) [inline WORKSPACE]: You can remove this. This is not used by any container_image target.

File: WORKSPACE [hunk 20]
```diff
@@ -56,26 +56,79 @@ new_container_pull(
...
+    name = "new_k8s_pause_arm64",
+    architecture = "arm64",
+    registry = "k8s.gcr.io",
+    repository = "pause",
+    tag = "3.1",
+)
+
+new_container_pull(
```
Comment (xingao267) [inline WORKSPACE]: Can remove this as well.

File: WORKSPACE [hunk 21]
```diff
@@ -56,26 +56,79 @@ new_container_pull(
...
+    name = "new_alpine_linux_ppc64le_tar",
+    architecture = "ppc64le",
+    format = "docker",
+    registry = "index.docker.io",
+    repository = "library/alpine",
+    tag = "3.8",
+)
+
+new_container_pull(
+    name = "new_k8s_pause_arm64",
```
Comment (xingao267) [inline WORKSPACE]: Can remove this as well.

File: WORKSPACE [hunk 22]
```diff
@@ -56,26 +56,79 @@ new_container_pull(
...
     name = "new_distroless_base",
     registry = "gcr.io",
     repository = "distroless/base",
 )
 
 new_container_pull(
-    name = "new_alpine_linux_armv6_tar",
-    architecture = "arm",
-    cpu_variant = "v6",
+    name = "new_distroless_cc",
```
Comment (xingao267) [inline WORKSPACE]: Can remove this as well.

**Pull request merged**